File: skills/vector-index/hooks/user_prompt_submit.py
#!/usr/bin/env python3
"""
UserPromptSubmit hook — recall prior resolutions for the incoming intent.

Claude Code (and compatible harnesses) invoke this with the hook payload on stdin
(session_id, transcript_path, cwd, prompt). It:

  1. Does a FAST, model-free recall (SQLite exact match + BM25) of intents similar
     to the prompt, scoped to the cwd's project with a global fallback, and prints
     a compact knowledge block to stdout — which the harness adds to the context
     BEFORE the assistant replies.
  2. Fires a DETACHED writer (`vindex intent record --async`) to embed and store
     the new intent, so the slow model load never blocks the prompt.

It must never break a turn: it honours VINDEX_INTENT_DISABLE, has a hard wall-
clock budget, and always exits 0 — any error is swallowed silently.
"""

import json
import os
import subprocess
import sys
import threading
from pathlib import Path

SCRIPTS = Path(__file__).resolve().parent.parent / "scripts"
sys.path.insert(0, str(SCRIPTS))

_TRUE = {"1", "true", "yes", "on"}


def _disabled() -> bool:
    return os.environ.get("VINDEX_INTENT_DISABLE", "").strip().lower() in _TRUE


def _read_payload() -> dict:
    try:
        return json.loads(sys.stdin.read() or "{}")
    except Exception:
        return {}


def _recall(prompt: str, cwd: str) -> str:
    # Stdlib-only path: `intents` does NOT import the model stack, and
    # light_project resolves the project without it either — so recall stays fast.
    import intents
    project = intents.light_project(cwd or None)
    store = intents.IntentStore.open()
    try:
        allow_embed = os.environ.get("VINDEX_INTENT_SYNC_EMBED", "").strip().lower() in _TRUE
        res = store.recall(prompt, project=project, allow_embed=allow_embed)
        return res.get("injection", "")
    finally:
        store.close()


def _fire_writer(prompt: str, cwd: str, session: str) -> None:
    """Detached: record + embed the new intent without blocking the hook. The
    child resolves the project itself (full resolver) from its cwd."""
    py = sys.executable
    cmd = [py, str(SCRIPTS / "vindex.py"), "intent", "record", "--async"]
    if session:
        cmd += ["--session", session]
    cmd += ["--", prompt]  # `--` so a prompt starting with '-' isn't read as a flag
    try:
        subprocess.Popen(
            cmd, stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
            start_new_session=True, cwd=cwd or None)
    except Exception:
        pass


def main() -> int:
    if _disabled():
        return 0
    payload = _read_payload()
    prompt = (payload.get("prompt") or "").strip()
    cwd = payload.get("cwd") or os.getcwd()
    session = payload.get("session_id") or ""
    if not prompt:
        return 0

    # Recall under a hard time budget so a cold sidecar load can never hang a turn.
    box: dict[str, str] = {}

    def work():
        try:
            box["out"] = _recall(prompt, cwd)
        except Exception:
            box["out"] = ""

    t = threading.Thread(target=work, daemon=True)
    t.start()
    t.join(timeout=float(os.environ.get("VINDEX_INTENT_TIMEOUT", "1.5")))
    injection = box.get("out", "")
    if injection:
        print(injection)

    # Always kick off the async writer (its own process, survives hook exit).
    try:
        _fire_writer(prompt, cwd, session)
    except Exception:
        pass
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception:
        sys.exit(0)

File: hooks/user_prompt_submit.ts
#!/usr/bin/env bun
/**
 * Pre-prompt hook — recall prior resolutions for the incoming intent.
 * Wired as `UserPromptSubmit` (Claude Code, Codex) and `BeforeAgent` (Gemini
 * CLI / Antigravity); all three deliver the same `{prompt, cwd, session_id}`.
 *
 * Reads the hook payload on stdin, does a FAST, model-free recall scoped to the
 * cwd's project, and prints a compact knowledge block to stdout as
 * `hookSpecificOutput.additionalContext` — the one injection format all three
 * harnesses accept (Gemini ignores non-JSON stdout entirely). Then fires a
 * DETACHED `vindex intent record` so storage never blocks the turn. Honours
 * VINDEX_INTENT_DISABLE, has a wall-clock budget, and always exits 0.
 */
import { spawn } from 'node:child_process'
import { fileURLToPath } from 'node:url'
import { dirname, join } from 'node:path'
import { IntentStore } from '../src/intents/store.ts'
import type { RecallMatch } from '../src/intents/store.ts'
import { INTENT_DISABLED, INTENT_MAX_TOKENS } from '../src/config.ts'
import { isMachineText } from '../src/transcript.ts'


const HERE = dirname(fileURLToPath(import.meta.url))
const CLI  = join(HERE, '..', 'src', 'cli', 'index.ts')

async function readStdin (): Promise<Record<string, unknown>> {
  try {
    const chunks: Buffer[] = []
    for await (const c of process.stdin)
      chunks.push(c as Buffer)
    return JSON.parse(Buffer.concat(chunks).toString('utf8') || '{}')
  }
  catch {
    return {}
  }
}

/**
 * Gemini HTML-escapes `<` and `>` inside additionalContext, which would turn
 * the XML delimiters into `&lt;vindex-intent-memory&gt;`, so BeforeAgent gets a
 * markdown heading instead. Everywhere else keeps the tag pair.
 */
function delimiters (event: string): [string, string] {
  return event === 'BeforeAgent'
    ? [ '### vindex intent memory', '### end vindex intent memory' ]
    : [ '<vindex-intent-memory>', '</vindex-intent-memory>' ]
}

function formatInjection (matches: RecallMatch[], event: string): string {
  if (!matches.length)
    return ''

  const [ open, close ] = delimiters(event)
  const lines           = [ open, 'Prior resolutions for similar intents:' ]
  let budget            = INTENT_MAX_TOKENS * 4 // ~chars
  for (const m of matches) {
    const exc  = (m.response_excerpt ?? '').replace(/\s+/g, ' ').trim()
    const line = `- [${m.outcome} ${m.score.toFixed(2)}] ${m.intent}${exc ? ` → ${exc}` : ''}`
    if (line.length > budget)
      break
    budget -= line.length
    lines.push(line)
  }
  lines.push(close)
  return lines.join('\n')
}

/**
 * The portable hook response. Every harness reads
 * `hookSpecificOutput.additionalContext`; `{}` is the valid no-op (Gemini
 * requires a JSON object on stdout either way).
 */
function emit (event: string, context: string): void {
  const payload = context
    ? { hookSpecificOutput: { hookEventName: event, additionalContext: context }}
    : {}
  process.stdout.write(JSON.stringify(payload) + '\n')
}

function fireWriter (prompt: string, cwd: string, session: string): void {
  try {
    const args = [ 'run', CLI, 'intent', 'record' ]
    if (session)
      args.push('--session', session)
    args.push('--', prompt)

    const child = spawn('bun', args, {
      cwd:      cwd || undefined,
      detached: true,
      stdio:    'ignore',
      env:      process.env,
    })
    child.unref()
  }
  catch { /* never break the turn */ }
}

async function main (): Promise<void> {
  if (INTENT_DISABLED)
    return emit('UserPromptSubmit', '')

  const payload = await readStdin()
  const event   = typeof payload.hook_event_name === 'string' ? payload.hook_event_name : 'UserPromptSubmit'
  const prompt  = String(payload.prompt ?? '').trim()
  const cwd     = typeof payload.cwd === 'string' ? payload.cwd : process.cwd()
  const session = typeof payload.session_id === 'string' ? payload.session_id : ''
  if (!prompt || isMachineText(prompt))
    return emit(event, '')

  const budgetMs = Number(process.env.VINDEX_INTENT_TIMEOUT || '1.5') * 1000
  const recall   = (async () => {
    const store = new IntentStore()
    try {
      return await store.recall(prompt, '', 3)
    }
    catch {
      return [] as RecallMatch[]
    }
  })()
  const timeout = new Promise<RecallMatch[]>(r => setTimeout(() => r([]), budgetMs))
  const matches = await Promise.race([ recall, timeout ])

  emit(event, formatInjection(matches, event))
  fireWriter(prompt, cwd, session)
}

main().then(() => process.exit(0))
  .catch(() => process.exit(0))
